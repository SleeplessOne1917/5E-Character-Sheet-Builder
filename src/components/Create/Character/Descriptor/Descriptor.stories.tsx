import { ComponentMeta, ComponentStory } from '@storybook/react';
import Descriptor, { DescriptorProps } from './Descriptor';

export default {
	title: 'Components/Create/Character/Descriptor',
	component: Descriptor,
	argTypes: { toggleOpen: { type: 'function' } }
} as ComponentMeta<typeof Descriptor>;

const Template: ComponentStory<typeof Descriptor> = (args: DescriptorProps) => (
	<Descriptor {...args} />
);

export const OneParagraph = Template.bind({});
OneParagraph.args = {
	title: 'Bar',
	description: 'This is a paragraph.'
};

export const MultipleParagraphs = Template.bind({});
MultipleParagraphs.args = {
	title: 'Baz',
	description: ['This is a paragraph.', 'This is another paragraph.']
};
