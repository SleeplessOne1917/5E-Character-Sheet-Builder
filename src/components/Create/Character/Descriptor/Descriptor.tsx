import { ChevronDownIcon, ChevronUpIcon } from '@heroicons/react/24/solid';
import {
	KeyboardEventHandler,
	forwardRef,
	memo,
	useCallback,
	useState,
	ReactNode
} from 'react';

import classes from './Descriptor.module.css';
import { handleKeyDownEvent } from '../../../../services/handlerService';
import dynamic from 'next/dynamic';
import LoadingSpinner from '../../../LoadingSpinner/LoadingSpinner';
const MarkdownParser = dynamic(
	() => import('../../../MarkdownParser/MarkdownParser'),
	{
		loading: () => (
			<div
				style={{
					display: 'flex',
					alignItems: 'center',
					justifyContent: 'center'
				}}
			>
				<LoadingSpinner />
			</div>
		)
	}
);

export type DescriptorProps = {
	title: string;
	description: string | string[];
	table?: ReactNode;
};

const Descriptor = forwardRef<HTMLDivElement, DescriptorProps>(
	({ title, description, table }: DescriptorProps, ref) => {
		const [isOpen, setIsOpen] = useState(false);

		const toggleOpen = useCallback(
			() => setIsOpen(prevState => !prevState),
			[setIsOpen]
		);

		const toggleOpenKeyDown: KeyboardEventHandler<HTMLDivElement> = useCallback(
			event => {
				handleKeyDownEvent<HTMLDivElement>(event, toggleOpen);
			},
			[toggleOpen]
		);

		return (
			<div className={classes.descriptor} data-testid="descriptor">
				<div
					className={classes.title}
					tabIndex={0}
					aria-label={title}
					onClick={toggleOpen}
					onKeyDown={toggleOpenKeyDown}
					style={{
						borderBottomLeftRadius: isOpen ? '0' : '0.5rem',
						borderBottomRightRadius: isOpen ? '0' : '0.5rem'
					}}
					role="button"
					ref={ref}
				>
					{title}
					{isOpen ? (
						<ChevronUpIcon className={classes.chevron} />
					) : (
						<ChevronDownIcon className={classes.chevron} />
					)}
				</div>
				{isOpen && (
					<div className={classes.content}>
						{table && <div className={classes['table-container']}>{table}</div>}
						<MarkdownParser input={description} />
					</div>
				)}
			</div>
		);
	}
);

Descriptor.displayName = 'Descriptor';

export default memo(Descriptor);
