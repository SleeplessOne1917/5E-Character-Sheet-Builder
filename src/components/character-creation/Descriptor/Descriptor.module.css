.descriptor {
    width: 70%;
    margin: 1rem 0;
}

.title {
    background-color: var(--davy);
    padding: 1rem;
    cursor: pointer;
    display: flex;
    align-items: center;
    justify-content: space-between;
    font-size: 1.7rem;
    color: var(--heaven);
    font-family: var(--font-fantasy);
    border-top-left-radius: 0.75rem;
    border-top-right-radius: 0.75rem;
}

.title:focus {
    outline: var(--focus-outline);
}

.content {
    background-color: var(--davy-light);
    height: fit-content;
    padding: 1rem;
    color: var(--heaven-light);
    border-bottom-left-radius: 0.25rem;
    border-bottom-right-radius: 0.25rem;
}

.chevron {
    height: 3rem;
    width: 3rem;
}

.content p:first-child {
    margin-block-start: 0;
}