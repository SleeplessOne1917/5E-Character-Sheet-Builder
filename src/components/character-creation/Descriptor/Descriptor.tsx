import { ChevronDownIcon, ChevronUpIcon } from '@heroicons/react/solid';
import {
	KeyboardEventHandler,
	forwardRef,
	memo,
	useCallback,
	useState,
	ReactNode
} from 'react';

import classes from './Descriptor.module.css';
import { handleKeyDownEvent } from '../../../services/handlerService';

export type DescriptorProps = {
	title: string;
	description: string | string[];
	table?: ReactNode;
};

const Descriptor = forwardRef<HTMLDivElement, DescriptorProps>(
	({ title, description, table }: DescriptorProps, ref) => {
		const [isOpen, setIsOpen] = useState(false);

		const toggleOpen = useCallback(
			() => setIsOpen(prevState => !prevState),
			[setIsOpen]
		);

		const toggleOpenKeyDown: KeyboardEventHandler<HTMLDivElement> = useCallback(
			event => {
				handleKeyDownEvent<HTMLDivElement>(event, toggleOpen);
			},
			[toggleOpen]
		);

		return (
			<div className={classes.descriptor} data-testid="descriptor">
				<div
					className={classes.title}
					tabIndex={0}
					aria-label={title}
					onClick={toggleOpen}
					onKeyDown={toggleOpenKeyDown}
					style={{
						borderBottomLeftRadius: isOpen ? '0' : '0.5rem',
						borderBottomRightRadius: isOpen ? '0' : '0.5rem'
					}}
					role="button"
					ref={ref}
				>
					{title}
					{isOpen ? (
						<ChevronUpIcon className={classes.chevron} />
					) : (
						<ChevronDownIcon className={classes.chevron} />
					)}
				</div>
				{isOpen && (
					<div className={classes.content}>
						{table && <div className={classes['table-container']}>{table}</div>}
						{Array.isArray(description) ? (
							description.map((d, i) => <p key={i}>{d}</p>)
						) : (
							<p>{description}</p>
						)}
					</div>
				)}
			</div>
		);
	}
);

Descriptor.displayName = 'Descriptor';

export default memo(Descriptor);
