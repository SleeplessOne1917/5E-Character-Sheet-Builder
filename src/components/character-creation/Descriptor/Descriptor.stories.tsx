import { ComponentMeta, ComponentStory } from '@storybook/react';
import Descriptor, { DescriptorProps } from './Descriptor';

export default {
	title: 'Components/Descriptor',
	component: Descriptor,
	argTypes: { toggleOpen: { type: 'function' } }
} as ComponentMeta<typeof Descriptor>;

const Template: ComponentStory<typeof Descriptor> = (args: DescriptorProps) => (
	<Descriptor {...args} />
);

export const Closed = Template.bind({});
Closed.args = {
	isOpen: false,
	title: 'Foo',
	description: 'text'
};

export const OneParagraph = Template.bind({});
OneParagraph.args = {
	isOpen: true,
	title: 'Bar',
	description: 'This is a paragraph.'
};

export const MultipleParagraphs = Template.bind({});
MultipleParagraphs.args = {
	isOpen: true,
	title: 'Baz',
	description: ['This is a paragraph.', 'This is another paragraph.']
};
