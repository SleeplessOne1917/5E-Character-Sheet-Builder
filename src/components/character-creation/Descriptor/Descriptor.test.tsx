import '@testing-library/jest-dom';

import * as stories from './Descriptor.stories';

import { render, screen } from '@testing-library/react';

import { composeStories } from '@storybook/testing-react';
import userEvent from '@testing-library/user-event';

const { Closed, MultipleParagraphs, OneParagraph } = composeStories(stories);

describe('renders correctly', () => {
	it('when closed', () => {
		render(<Closed />);

		expect(screen.getByTestId(/descriptor/i)).toMatchSnapshot();
	});

	it('when open with single paragraph', () => {
		render(<OneParagraph />);

		expect(screen.getByTestId(/descriptor/i)).toMatchSnapshot();
	});

	it('when open with multiple paragraphs', () => {
		render(<MultipleParagraphs />);

		expect(screen.getByTestId(/descriptor/i)).toMatchSnapshot();
	});
});

describe('calls toggleOpen', () => {
	it('on click', async () => {
		const toggleOpenMock = jest.fn();

		render(<Closed toggleOpen={toggleOpenMock} />);

		await userEvent.click(screen.getByRole('button'));

		expect(toggleOpenMock).toHaveBeenCalled();
	});

	it('on enter keypress', async () => {
		const toggleOpenMock = jest.fn();

		render(<Closed toggleOpen={toggleOpenMock} />);

		await userEvent.tab();
		await userEvent.keyboard('{Enter}');

		expect(toggleOpenMock).toHaveBeenCalled();
	});

	it('on space keypress', async () => {
		const toggleOpenMock = jest.fn();

		render(<Closed toggleOpen={toggleOpenMock} />);

		await userEvent.tab();
		await userEvent.keyboard(' ');

		expect(toggleOpenMock).toHaveBeenCalled();
	});
});

describe('has expected styles', () => {
	it('when closed', () => {
		render(<Closed />);

		expect(screen.getByText(/Foo/i)).toHaveStyle({
			'border-bottom-left-radius': '0.5rem',
			'border-bottom-right-radius': '0.5rem'
		});
	});

	it('when open', () => {
		render(<OneParagraph />);

		expect(screen.getByText(/Bar/i)).toHaveStyle({
			'border-bottom-left-radius': '0',
			'border-bottom-right-radius': '0'
		});
	});
});
