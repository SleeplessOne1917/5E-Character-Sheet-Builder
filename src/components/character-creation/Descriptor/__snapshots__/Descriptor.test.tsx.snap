// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when open with multiple paragraphs 1`] = `
<div
  class="descriptor"
  data-testid="descriptor"
>
  <div
    aria-label="Baz"
    class="title"
    role="button"
    style="border-bottom-left-radius: 0.5rem; border-bottom-right-radius: 0.5rem;"
    tabindex="0"
  >
    Baz
    <svg
      aria-hidden="true"
      class="chevron"
      fill="currentColor"
      viewBox="0 0 20 20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z"
        fill-rule="evenodd"
      />
    </svg>
  </div>
</div>
`;

exports[`renders correctly when open with single paragraph 1`] = `
<div
  class="descriptor"
  data-testid="descriptor"
>
  <div
    aria-label="Bar"
    class="title"
    role="button"
    style="border-bottom-left-radius: 0.5rem; border-bottom-right-radius: 0.5rem;"
    tabindex="0"
  >
    Bar
    <svg
      aria-hidden="true"
      class="chevron"
      fill="currentColor"
      viewBox="0 0 20 20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z"
        fill-rule="evenodd"
      />
    </svg>
  </div>
</div>
`;
